### ISSUE TITLE
CPNest cannot pickle and resume

### ISSUE THREAD

**Issue by GCArullo** [johnveitch/cpnest#87]
[Line 72 of NestedSamling.py](https://github.com/johnveitch/cpnest/commit/b9028a0c69500e25d2320ed4063e4af5fe347126#diff-7c291aa12dcb6bf87414dbc07ba0cf167d70eff205f35e9606a82d5e79a1bf2eR72) introduced in b9028a0c69500e25d2320ed4063e4af5fe347126 adds a lambda function, breaking pickling. In turn, this does not allow to resume interrupted cpnest jobs. 

This also prevents long-lived pyRing runs to be executed through condor, since condor frequently kills and restarts the jobs. More details below.

---

For reference, the issue with lambda functions and pickling is explained [here](https://stackoverflow.com/questions/72339545/attributeerror-cant-pickle-local-object-locals-lambda).

The problem can be reproduced by adding `resume=True` to this [example](https://github.com/johnveitch/cpnest/blob/master/examples/gaussianmixture.py#L44). I tested it with both py3.7 and py3.8. When starting the above example and killing the run during sampling, the following error arises:

```
MacBook-Pro-di-Gregorio:examples gregoriocarullo$ python gaussianmixture.py 
test_run (__main__.GaussianMixtureTestCase) ... 2022-11-30, 14:32:20 - cpnest.cpnest.CPNest                  : Running with 6 parallel threads
CPNEST: populate samplers: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1024/1024 [00:32<00:00, 31.85it/s]
                                                                                                                                                                                                                                         
^C2022-11-30, 14:33:01 - cpnest.cpnest                         : Handling signal 2                                                                                                                                                       
2022-11-30, 14:33:01 - cpnest.NestedSampling.NestedSampler   : Checkpointing nested sampling                                                                                                                                             
2022-11-30, 14:33:01 - cpnest.cpnest                         : Handling signal 2                                                                                                                                                         
2022-11-30, 14:33:01 - cpnest.cpnest                         : Handling signal 2                                                                                                                                                         
2022-11-30, 14:33:01 - cpnest.cpnest                         : Handling signal 2█████████████████████████████████████████████████████████████████████████████████████████████████████████████████▌  | 1009/1024 [00:32<00:00, 129.40it/s]
2022-11-30, 14:33:01 - cpnest.cpnest                         : Handling signal 2████████████████████████████████▎                                                                                     | 502/1000 [00:19<00:10, 49.41it/s]
2022-11-30, 14:33:01 - cpnest.cpnest                         : Handling signal 2████████████▉                                                                                                         | 389/1000 [00:14<00:10, 56.70it/s]
2022-11-30, 14:33:01 - cpnest.sampler.Sampler                : Checkpoint excepted in sampler██████████████████████████████████████████████████████████████████████████████████████████████████████▎ | 990/1000 [00:20<00:00, 172.32it/s]
2022-11-30, 14:33:01 - cpnest.cpnest                         : Handling signal 2
2022-11-30, 14:33:01 - cpnest.sampler.Sampler                : Checkpoint excepted in sampler
2022-11-30, 14:33:01 - cpnest.sampler.Sampler                : Checkpoint excepted in sampler
2022-11-30, 14:33:01 - cpnest.sampler.Sampler                : Checkpoint excepted in sampler
2022-11-30, 14:33:01 - cpnest.sampler.Sampler                : Checkpoint excepted in sampler
2022-11-30, 14:33:01 - cpnest.sampler.Sampler                : Checkpoint excepted in sampler
ERROR

======================================================================
ERROR: test_run (__main__.GaussianMixtureTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/cpnest-0.11.4-py3.7-macosx-10.13-x86_64.egg/cpnest/NestedSampling.py", line 452, in nested_sampling_loop
    self.consume_sample()
  File "/usr/local/lib/python3.7/site-packages/cpnest-0.11.4-py3.7-macosx-10.13-x86_64.egg/cpnest/NestedSampling.py", line 342, in consume_sample
    acceptance, sub_acceptance, self.jumps, proposed = self.manager.consumer_pipes[self.queue_counter].recv()
  File "/usr/local/Cellar/python/3.7.4_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/multiprocessing/connection.py", line 250, in recv
    buf = self._recv_bytes()
  File "/usr/local/Cellar/python/3.7.4_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/multiprocessing/connection.py", line 407, in _recv_bytes
    buf = self._recv(4)
  File "/usr/local/Cellar/python/3.7.4_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/multiprocessing/connection.py", line 379, in _recv
    chunk = read(handle, remaining)
  File "/usr/local/lib/python3.7/site-packages/cpnest-0.11.4-py3.7-macosx-10.13-x86_64.egg/cpnest/cpnest.py", line 30, in sighandler
    raise CheckPoint()
cpnest.cpnest.CheckPoint

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "gaussianmixture.py", line 47, in test_run
    self.work.run()
  File "/usr/local/lib/python3.7/site-packages/cpnest-0.11.4-py3.7-macosx-10.13-x86_64.egg/cpnest/cpnest.py", line 283, in run
    self.NS.nested_sampling_loop()
  File "/usr/local/lib/python3.7/site-packages/cpnest-0.11.4-py3.7-macosx-10.13-x86_64.egg/cpnest/NestedSampling.py", line 461, in nested_sampling_loop
    self.checkpoint()
  File "/usr/local/lib/python3.7/site-packages/cpnest-0.11.4-py3.7-macosx-10.13-x86_64.egg/cpnest/NestedSampling.py", line 498, in checkpoint
    pickle.dump(self, f)
AttributeError: Can't pickle local object 'OrderedLivePoints.__init__.<locals>.<lambda>'
```

Subsequent resuming then fails with:

```
MacBook-Pro-di-Gregorio:examples gregoriocarullo$ python gaussianmixture.py 
test_run (__main__.GaussianMixtureTestCase) ... 2022-11-30, 14:33:08 - cpnest.cpnest.CPNest                  : Running with 6 parallel threads
ERROR

======================================================================
ERROR: test_run (__main__.GaussianMixtureTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "gaussianmixture.py", line 44, in setUp
    self.work=cpnest.CPNest(GaussianMixtureModel(),verbose=2,nthreads=6,nlive=1024,maxmcmc=1000,poolsize=1000,nslice=6, resume=True)
  File "/usr/local/lib/python3.7/site-packages/cpnest-0.11.4-py3.7-macosx-10.13-x86_64.egg/cpnest/cpnest.py", line 193, in __init__
    self.user)
  File "/usr/local/lib/python3.7/site-packages/cpnest-0.11.4-py3.7-macosx-10.13-x86_64.egg/cpnest/NestedSampling.py", line 507, in resume
    obj = pickle.load(f)
EOFError: Ran out of input

----------------------------------------------------------------------
Ran 1 test in 0.423s

FAILED (errors=1)
```

**Comment by GCArullo** [comment]
@mj-will could you please have a look?

**Comment by mj-will** [comment]
@GCArullo sure, I'll try to take a look some point this week.

**Comment by johnveitch** [comment]
@GCArullo Does this commit on current master fix the issue for you?

- 1562193825d8c420ea9c4abb9ec3bf6bf7471711

**Comment by GCArullo** [comment]
Yes, it does! Thank you very much John. 

The issue can be closed (it did not work when I tried to close it myself, not sure why).

**Comment by johnveitch** [comment]
I'll keep it open until this makes it into a bugfix release

**Comment by johnveitch** [comment]
Fixed in v0.11.5
